ITE/4/README.adoc

```diff
@@ -0,0 +1,510 @@
+commit as a product.
+
+===== in-toto link attestation for GitHub Actions building from a merge commit
+
+GitHub Actions can be used to set up a workflow for continuous integration (CI).
+Workflows can be triggered on push and an attestation can be generated for the
+resulting build and CI report.
+
+
+```
+{
+    "_type": "link",
+    "name": "github-actions-build-pull-request-250",
+    "command": "",
+    "materials": {
+        "github+https://github.com/in-toto/in-toto/commit/f1c5d201887e226cadac5792a203ac3eae347add": <HASH>
+    },
+    "products": {
+        "github+https://github.com/in-toto/in-toto/commit/f1c5d201887e226cadac5792a203ac3eae347add/checks?check_suite_id=<ID>": <HASH>
+    },
+    "byproducts": {
+        "stderr": "",
+        "stdout": "",
+        "return-value": ""
+    },
+    "environment": {
+        "variables": "",
+        "filesystem": "",
+        "workdir": ""
+    }
+}
+```
+
+[[reasoning]]
+== Reasoning
+
+This ITE proposes a change in the URI scheme of artifacts in in-toto metadata.
+The proposed change closely matches the URI structure laid out in RFC 3986. The
+URI structure proposed in that document is widely accepted and is versatile
+enough to allow for a wide variety of references. It is also easy to implement
+and extend support for due to the presence of a large number of standard
+libraries.
+
+[[backwards-compatibility]]
+== Backwards Compatibility
+
+If in-toto metadata is generated using an implementation of in-toto conforming
+to this ITE, verification using a non-conforming implementation can fail. It is
+possible for the verification workflow to progress if the inspections don't
+use generic URIs.
+
+However, a conforming implementation SHOULD be capable of verifying in-toto
+metadata generated using a non-conforming implementation, as an ITE-4 conforming
+system MUST also conform to the actual in-toto specification.
+
+It's also possible two conforming systems may be unable to verify the other's
+in-toto metadata as they're unaware of how to resolve certain URI tokens used by
+the other. This is again because of the possibility of inspections containing
+URIs that the other system is unable to resolve and calculate cryptographic
+information for.
+
+[[security]]
+== Security
+
+As per the in-toto specification, the only direct interaction in-toto tools have
+with artifacts is to record their hashes using one or more cryptographic hash
+algorithms. Other artifact operations such as verifying the artifact rules
+rely on the hashes recorded in the link metadata.
+
+We found that the recording of hashes of abstract resources, some of which may
+live at remote locations, is more complicated than the recording of hashes of
+artifacts in the local filesystem. It is, therefore, important for ITE-4
+compliant in-toto tools to handle the resolving of generic URIs to these
+abstract resources securely. It is quite possible that the contents of these
```

Comment:
What does "securely" mean here?  What if a resource is unavailable?

Reply:
I meant that they must be careful with how they serialize. Additionally, do you think it's okay to add that implementers must handle situations where resources are unavailable by failing appropriately, as the context demands?

Reply:
Okay, can you clarify in the text please?

Reply:
I've updated it to read since I think the exact handling of a missing resource may be context dependent:

> Implementors must also take care to ensure they handle situations where a resource may be unavailable, and fail appropriately.

Do you think we need to be more strict and say the workflow must fail? I was also thinking of inspections that explicitly check if a resource is unavailable to pass.

Reply:
CC: @SantiagoTorres 

I'm not sure what makes sense.  Should every in-toto implementation make its own decision?  Should the layout specify what should happen in this case?  Should the user be prompted (whatever that means)?

Reply:
The workflow I was imagining/hoping-for was to have resolvers exist like plug-ins for implementations. I think we should `MUST` fail if it needs to resolve a resource but doesn't have the code to do so makes sense to me.

I didn't think of having information in the layout to say what should happen if that's the case. I think it may be useful but I wonder if it would make the layout too verbose...

Reply:
I support clearly defined behavior such as "MUST fail" and not making the layout more verbose.

I also think it's worth to stress once more in this section that the availability of resources is *only* relevant for resources that are hashed and recoded in the course of running inspections, right?

Also note that the reference implementation currently does not allow to parametrize which materials and products should be hashed and recorded for inspections, instead we record all files in the verification cwd.

Reply:
> I think we should MUST fail if it needs to resolve a resource but doesn't have the code to do so makes sense to me.

I agree with this, but this section is talking about compliant tools, meaning they presumably have the means to resolve the URI but the resource is missing at that location. That, I believe, should be handled per context, per type of resource.

Reply:
I think we can rephrase and clarify. I can see how many readers could read what Justin read from what is there.

Reply:
I think https://github.com/in-toto/ITE/pull/6/commits/de11c18cac1fc07be8f4013422694c359c369deb clarifies that.